rats-devtools/src/python/rats/docs/_app.py

```diff
@@ -0,0 +1,37 @@
+import click
+
+from rats import apps, devtools  # type: ignore[reportAttributeAccessIssue]
+
+from ._commands import RatsCiCommands
+from ._ids import PluginServices
+
+
+class RatsDocsPlugin(apps.AnnotatedContainer):
+    _app: apps.Container
+
+    def __init__(self, app: apps.Container) -> None:
+        self._app = app
+
+    @apps.group(devtools.AppServices.GROUPS.CLI_ROOT_COMMANDS)
+    def docs_command(self) -> click.Command:
+        provider = devtools.CommandProvider(
+            command_names=frozenset(
+                [
+                    "mkdocs-serve",
+                    "mkdocs-build",
+                    "sphinx-build",
+                    "build-jupytext-notebooks",
+                ]
+            ),
+            service_mapper=lambda name: PluginServices.GROUPS.command(name),
+            container=self,
+        )
```

Comment:
I am not happy with the need to list the commands here.
Shouldn't it be enough to declare the command in the container?

You are currently coupling the implementation of this method with both the existence of these commands and with the strings passed to `PluginServices.GROUPS.command` when they are declared as services.

If instead there was a decorator that would register the command and its name to a group, then we could provide all the information about a command when we declared it, and this method would just need the group id.

Reply:
i want to add the command annotations back in another pr. this is my attempt to define the cli with lazy loading support, which mostly requires a way to get the list of command names, and translate the command name to a click command class. a non-lazy version of this could probably just take a set of command classes and extract the names automatically.

i can follow this up with a pr to use the annotations package to create some shortcuts for classes with multiple commands

Reply:
I took a swing at inverting this dependency in #161, by introducing groups that look like `@apps.group(command_tree.CommandTreeServices.GROUPS.subcommands("rats-devtools ci"))`, then collecting that group in the parent service provider

@ms-lolo, things are pretty fast when running this with either of the two executables enabled. What did you see taking up a bunch of time while you were testing the design? Want to make sure I didn't trip over something I'm not aware of

Reply:
the requirement i made for myself was to have the library avoid executing any code the user would own, until i know one of those commands is the one being executed. it's not as good as avoiding the import completely, since you mentioned that would most likely be the slowest part (all the plugin entrypoints logic), but the next best thing i could think of was to avoid the user being able to slow everything down accidentally.

```python
@apps.service(PluginServices.command("mkdocs-serve"))
def mkdocs_serve_command(self) -> click.Command:
    @click.command
    def mkdocs_serve() -> None:
        """Combine our docs across components and run the mkdocs serve command."""
        self._do_mkdocs_things("serve")
```

figured we can have a test validate that the provider here doesn't get called unless we are in the `docs` sub-command.